### ISSUE TITLE
[HOLD for payment 2023-09-18] [$1000] Bank account - Inconsistency bug: Checking ToS box before filling fields throws error in mweb but not in android app

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#26009]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login to staging new dot
2. Go to workspaces and select Bank account
3. Click on connect manually and check the ToS box before filling the routing and bank account number fields
4. Observe that an error is thrown
5. Login to android app and repeat steps 2-3
6. Observe that no error is thrown when checking and unchecking ToS box before filling routing and bank account number fields

## Expected Result:
Error throwing should be consistent across platforms

## Actual Result:
Inconsistent error throwing

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.57-5

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/acd3ef2a-c62e-4a8a-9367-cf45176e657f

https://github.com/Expensify/App/assets/78819774/4de07e4c-db19-49d4-96a1-2aa58d261d78

**Expensify/Expensify Issue URL:**

**Issue reported by:** @sofoniasN

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1692271454788889

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01feabe39f1c833087</li>
        <li>Upwork Job ID: 1696317266455986176</li>
        <li>Last Price Increase: 2023-08-29</li>
<li>Automatic offers: </li>
<ul>
<li>tamdao | Contributor | 26491662</li>
<li>sofoniasN | Reporter | 26491663</li>
</ul></li>
    </ul>
</details>

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Checking the ToS box before filling fields throws an error in MWeb but not in android app

### What is the root cause of that problem?
On mobile web, checking the ToS box causes the blur event for the text input field to fire which triggers validation. On native platforms, this is not the case.

### What changes do you think we should make in order to solve the problem?
Either of these solutions has to be selected for implementation.

**Solution 1**
On mobile web, add a preventDefault call for the checkbox event and then programmatically change the value of the checkbox using a state variable.

**Solution 2**
On native platforms, blur the text input when the checkbox is tapped or clicked.

### What alternative solutions did you explore? (Optional)
None.

**Comment by tamdao** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Bank account - Inconsistency bug: Checking ToS box before filling fields throws error in mweb but not in android app

### What is the root cause of that problem?
The root cause of this issue is `TextInput.autoFocus` prop

https://github.com/Expensify/App/blob/c029dc509d03fd13ece058ba89ab25c4318a9050/src/pages/ReimbursementAccount/BankAccountManualStep.js#L97-L98

The `Form` component you have logic the show the error message when the value change, but the error message only shows when the input already added into the `touchedInputs` list

https://github.com/Expensify/App/blob/c029dc509d03fd13ece058ba89ab25c4318a9050/src/components/Form.js#L150-L157

### What alternative solutions did you explore?
Update childrenWrapperWithProps logic to call `setTouchedInput` when the `child.props.autoFocus` is true

https://github.com/Expensify/App/blob/c029dc509d03fd13ece058ba89ab25c4318a9050/src/components/Form.js#L219-L357

 like this
```js
if (child.props.autoFocus) {
    setTouchedInput(child.props.inputID);
}
```


https://github.com/Expensify/App/assets/3086267/49340cc4-ade5-499c-8349-fcb71da64123

**Comment by tamdao** [comment]
Contributor details
Your Expensify account email: daotam4536@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0133d983376a3d0d3d

**Comment by strepanier03** [comment]
This was previously reported and the [GH](https://expensify.slack.com/archives/C049HHMV9SM/p1692271454788889) closed as a feature request, not a bug.

**Comment by strepanier03** [comment]
I'm going to move this to external so that we have consistency between devices. 

Note that on m/web AND android/native, the error happens if you enter an account number before the routing number. I think the easiest solution is to populate the error when the TOS checkbox or accounting number is entered before the routing number. That way both devices would be just as picky about the order.

**Comment by burczu** [comment]
@strepanier03 Just to be clear: based on [your comment](https://github.com/Expensify/App/issues/26009#issuecomment-1696596326), the correct behaviour is to show the validation error when we click/tap on the checkbox first (or trying to provide account number before routing number)? Ergo, we need to fix it on mobile, not on web. Am I right?

**Comment by burczu** [comment]
@tamdao Just wondering, what if there is more than one input with `autoFocus` property inside the Form? Probably it's rare case scenario, but I think we shouldn't add all of these inputs to the `touchedInput` list, but only this one that is really focused... WDYT?

**Comment by akinwale** [comment]
> @strepanier03 Just to be clear: based on [your comment](https://github.com/Expensify/App/issues/26009#issuecomment-1696596326), the correct behaviour is to show the validation error when we click/tap on the checkbox first (or trying to provide account number before routing number)? Ergo, we need to fix it on mobile, not on web. Am I right?

If this is the expected behaviour, Solution 2 in [my proposal](https://github.com/Expensify/App/issues/26009#issuecomment-1694568434) takes care of this.

**Comment by burczu** [comment]
@tamdao I agree no one should do that but, as far as I know, it's not programmatically forbidden anywhere, so in case there is more than one `autoFocus` all of them will be added to the `thouchedInput` array in your solution - that's why I think your proposal should contain some kind of check for it.

**Comment by tamdao** [comment]
@burczu if you want, I can add more logic to check that. But you know in development, we also need to compare the performance and the value of the logic to make sure that change valuable

**Comment by tamdao** [comment]
@burczu We can add a flag like this
```js
let isTouchedAutoFucusInput = false;
const childrenElements = React.Children.map(childNodes, (child) => {
....
    if (child.props.autoFocus && !isTouchedAutoFucusInput) {
       setTouchedInput(child.props.inputID);
       isTouchedAutoFucusInput = true;
     }
....
```

But I have tested the form with multiple autoFocus, it will cause validation issue when user open the screen like this
<img width="367" alt="Screenshot 2023-08-29 at 19 39 47" src="https://github.com/Expensify/App/assets/3086267/33305699-0eab-41bf-a2b2-12d7e0b4d734">
This issue happens because I add 2 autoFucus into 2 input and the second will show the error message

**Comment by burczu** [comment]
Thanks @tamdao for this investigation! So it seems this check is not necessary because having more than one `autoFocus` in the Form will cause unexpected behaviour anyway.

Let's wait for the confirmation from @strepanier03 about my concern [here](https://github.com/Expensify/App/issues/26009#issuecomment-1697191813). But once confirmed I think we will proceed with your proposal.

**Comment by strepanier03** [comment]
Sorry for the confusion and thanks for the ping to clarify. 

To bring consistency, the update needed here is only on Android/native so that when the TOS checkbox is checked before the routing number is entered, the error populates.

**Comment by burczu** [comment]
Thanks @strepanier03!

So, I think [the proposal](https://github.com/Expensify/App/issues/26009#issuecomment-1694596626) from @tamdao is the one we should proceed with - it fixes the issue in the Form component, so we may avoid similar issues in the future.

The solution 2 from @akinwale [proposal](https://github.com/Expensify/App/issues/26009#issuecomment-1694568434) is might also fix the issue but it's more like workaround that fixes the issue only in the bank account flow.

🎀 👀 🎀 C+ reviewed

**Comment by tamdao** [comment]
@burczu Sorry to distract you, but when I tested my solution, I found another issue. About the issue I found is focus the second input and check the checkbox the validation not show, because you know this input is not in the touched list. So here new proposal:

## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Bank account - Inconsistency bug: Checking ToS box before filling fields throws error in mweb but not in android app

### What is the root cause of that problem?
The root cause of this issue is value change for the input not trigger blur event. In this case the input is checkbox.

### What alternative solutions did you explore?
Update the `childrenWrapperWithProps` function in `Form` component

https://github.com/Expensify/App/blob/c029dc509d03fd13ece058ba89ab25c4318a9050/src/components/Form.js#L219-L357

Add new ref
```js
const focusedInput = useRef(null);
```

Update childrenWrapperWithProps logic to add `onFocus` function:

```js
... line 307
onFocus: (event) => {
    focusedInput.current = inputID;
    if (_.isFunction(child.props.onFocus)) {
        child.props.onFocus(event);
    }
},
onBlur: ....
```
and update the `onInputChange` function (line 336) to check the changed value is focused input or not

```js
const inputKey = key || inputID;

if (focusedInput.current && focusedInput.current !== inputKey) {
    setTouchedInput(focusedInput.current);
}
```

I have tested this change and it works well.

**Comment by burczu** [comment]
@tamdao I've just checked the new solution and it does the trick indeed - I like it. Please update your original proposal with the new one, and your last comment with something like:

```
## Proposal
[Updated](link to proposal)
```

Read more about it [here](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job).

**Comment by tamdao** [comment]
@burczu Thanks for your comment.

## Proposal
[Updated](https://github.com/Expensify/App/issues/26009#issuecomment-1694596626)

**Comment by puneetlath** [comment]
Hmm, interesting. @burczu @tamdao to make sure I understand, will this solution affect the behavior of any other types of form inputs? Or only check boxes?

**Comment by tamdao** [comment]
@puneetlath my solution fixes any inputs form not trigger blur event when user change value, not only checkbox

**Comment by tamdao** [comment]
@puneetlath short answer: No
Long answer: My change only adds logic the check changed input and the current focus input is different or not. If it's different, I will add the current focus input into touched list to let the validation check that input. And this change didn't change any old behavior

**Comment by puneetlath** [comment]
Ok, thanks for the explanation. Let's do it then.

**Comment by tamdao** [comment]
@burczu @puneetlath PR is ready https://github.com/Expensify/App/pull/26699. Please help me review it. Thanks ❤️

**Comment by burczu** [comment]
Just to inform: I'll be OOO on Thursday and Friday (7-8 September).

**Comment by tamdao** [comment]
That not true :(

**Comment by SofoniasN** [comment]
@puneetlath I noticed that the offer sent in upwork for the bug report is 50 dollars, but this issue was created before the new price scheme was announcement and is eligble for the original 250 payment . Could you please look into it. Thanks

**Comment by strepanier03** [comment]
### Payment summary:

- [x] External issue reporter - @SofoniasN - $250 paid via Upwork
- [x] Contributor that fixed the issue - @tamdao - $1000 paid via Upwork
- [ ] Contributor+ that helped on the issue and/or PR - @burczu - $1000 paid via Manual Requests

**Speed bonus assessment:** @tamdao assigned on 2023-09-04 / PR merged on 2023-09-08 - 4 days lapsed = not eligible

**Comment by strepanier03** [comment]
@burczu - Please submit your manual request for payment and let me know once you've done so. I will then assign Jason to issue payment. 

@SofoniasN - I have paid you via Upwork and since this was reported prior to our pay change I made sure it was $250. 

@tamdao - I took another look at the timeline based on your comment [here](https://github.com/Expensify/App/issues/26009#issuecomment-1712296691) but I confirmed the date you were assigned and the PR was merged was 4 days so I paid you out via Upwork at $1000.

**Comment by strepanier03** [comment]
@burczu - I totally spaced the checklist [here](https://github.com/Expensify/App/issues/26009#issuecomment-1714738174) so please take care of that and then submit for manual payment, sorry about that!

**Comment by burczu** [comment]
@strepanier03 I'm from Callstack, so I'm not eligible for payment for this.

**Comment by strepanier03** [comment]
Oh doh, I knew that. Thanks for reminding me @burczu 🤦‍♀️. 

I'll follow along and finish up once you have time to do the checklist. Thanks again!

**Comment by strepanier03** [comment]
I'm out of the office Monday-Tuesday next week. If urgent action is needed from a BZ either day, please reach out in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) for help. Otherwise I'll check in Wednesday.

**Comment by puneetlath** [comment]
Going to set this to weekly priority since only thing left is the checklist.

**Comment by puneetlath** [comment]
@burczu also, if you have too many issues to handle, please feel encouraged to ask in #contributor-plus if others want to take some over!

**Comment by burczu** [comment]
@puneetlath No worries, I've just managed to address all my other assignments and will start handling BZ checklists that are waiting for me.

**Comment by burczu** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

* [x]  [@burczu] The PR that introduced the bug has been identified. Link to the PR: This is not a regression - it just wasn't implemented this way.
* [x]  [@burczu] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: n/a
* [x]  [@burczu] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: n/a
* [x]  [@burczu] Determine if we should create a regression test for this bug.
* [x]  [@burczu] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
* [ ]  [@puneetlath / @strepanier03] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by puneetlath** [comment]
Thanks! I think we're good to go here.
